### PR TITLE
[luci-micro] Luci-micro optimizations: step 4

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This PR introduces emplace kernels, and adds some optimizations to remove unnecessary classes fields to achieve memory reduction.

issue: #9468

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci-micro/luci-interpreter/include/luci_interpreter/core/Tensor.h [hunk 1]
```diff
@@ -152,6 +152,8 @@ class Tensor
     _data = buffer;
   }
 
+  uint8_t *get_data_buffer() const { return _data; }
```
Comment (binarman) [inline compiler/luci-micro/luci-interpreter/include/luci_interpreter/core/Tensor.h:155]: Why not use `data()` method?

File: compiler/luci-micro/luci-interpreter/src/core/Kernel.h [hunk 2]
```diff
@@ -47,10 +47,17 @@ class Kernel
...
+  void setEmplaceValue(bool is_emplace) { _is_emplace = is_emplace; }
+  // Return _is_emplace value.
+  bool getEmplaceValue() const { return _is_emplace; }
+
 protected:
   // NOTE Prefer not to use these in derived classes.
   const std::vector<const Tensor *> _inputs;
   const std::vector<Tensor *> _outputs;
+
+  bool _is_emplace;
```
Comment (binarman) [inline compiler/luci-micro/luci-interpreter/src/core/Kernel.h:60]: maybe `_is_inplace`?

File: compiler/luci-micro/luci-interpreter/include/luci_interpreter/core/Tensor.h (hunk 1 above)
Comment (BalyshevArtem) [reply compiler/luci-micro/luci-interpreter/include/luci_interpreter/core/Tensor.h]: done

File: compiler/luci-micro/luci-interpreter/src/core/Kernel.h (hunk 2 above)
Comment (BalyshevArtem) [reply compiler/luci-micro/luci-interpreter/src/core/Kernel.h]: done, thanks :)

**Review by binarman — APPROVED** [review]
LGTM,
Also, think about CRTP "InplaceKernel" class, which could implement wrapper functionality around actual execute.

**Pull request merged**